File: src/main/java/com/kingmang/lazurite/parser/AST/Statements/ClassDeclarationStatement.java
package com.kingmang.lazurite.parser.AST.Statements;

import com.kingmang.lazurite.core.ClassDeclarations;
import com.kingmang.lazurite.parser.AST.Expressions.AssignmentExpression;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;

import java.util.ArrayList;
import java.util.List;

public final class ClassDeclarationStatement implements Statement {
    
    public final String name;
    public final List<FunctionDefineStatement> methods;
    public final List<AssignmentExpression> fields;
    
    public ClassDeclarationStatement(String name) {
        this.name = name;
        methods = new ArrayList<>();
        fields = new ArrayList<>();
    }
    
    public void addField(AssignmentExpression expr) {
        fields.add(expr);
    }

    public void addMethod(FunctionDefineStatement statement) {
        methods.add(statement);
    }

    @Override
    public void execute() {
        ClassDeclarations.set(name, this);
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return String.format("class %s {\n  %s  %s}", name, fields, methods);
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/ClassDeclarationStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.core.ClassDeclarations
import com.kingmang.lazurite.parser.ast.expressions.AssignmentExpression
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor

data class ClassDeclarationStatement(val name: String) : Statement {
    val methods: MutableList<FunctionDefineStatement> = ArrayList()
    val fields: MutableList<AssignmentExpression> = ArrayList()

    fun addField(expr: AssignmentExpression) {
        this.fields.add(expr)
    }

    fun addMethod(statement: FunctionDefineStatement) {
        this.methods.add(statement)
    }

    override fun execute() {
        ClassDeclarations.set(this.name, this)
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String =
        String.format("class %s {\n  %s  %s}", name, fields, methods)
}
